### PR TITLE
Account for Bad Blocks in Gossip

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Spec Update

**What does this PR do? Why is it needed?**

- [x] Add a bad block cache, to keep a record of all blocks which had invalid state transitions/signatures/etc. 
- [x] Reject blocks whose parent blocks are invalid.
- [x] Reject attestations whose voted block is invalid.
- [x] Reject Aggregate and Proof whose pointed block is invalid.
- [x] Add tests for all bad block checks.

**Which issues(s) does this PR fix?**

Part of #6716 and corresponding PR in the specs repo is
https://github.com/ethereum/eth2.0-specs/pull/1956

**Other notes for review**

### CODE REVIEW

File: beacon-chain/sync/pending_blocks_queue.go [hunk 1]
```diff
@@ -64,8 +64,31 @@ func (s *Service) processPendingBlocks(ctx context.Context) error {
...
+			// Set block as bad if its parent block is bad too.
+			if parentIsBad {
+				s.setBadBlock(blkRoot)
+			}
+			// Remove block from queue.
+			s.pendingQueueLock.Lock()
+			delete(s.slotToPendingBlocks, slot)
+			delete(s.seenPendingBlocks, blkRoot)
+			s.pendingQueueLock.Unlock()
+			span.End()
```
Comment (terencechain) [inline beacon-chain/sync/pending_blocks_queue.go:89]: Should the span end here? The line after is `continue`

File: beacon-chain/sync/validate_aggregate_proof.go [hunk 2]
```diff
@@ -54,6 +54,10 @@ func (s *Service) validateAggregateAndProof(ctx context.Context, pid peer.ID, ms
 	if s.hasSeenAggregatorIndexEpoch(m.Message.Aggregate.Data.Target.Epoch, m.Message.AggregatorIndex) {
 		return pubsub.ValidationIgnore
 	}
+	// Check that the block being voted on isn't invalid.
+	if s.hasBadBlock(bytesutil.ToBytes32(m.Message.Aggregate.Data.BeaconBlockRoot)) {
```
Comment (terencechain) [inline beacon-chain/sync/validate_aggregate_proof.go:58]: Should we check target root too?

File: beacon-chain/sync/pending_blocks_queue.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/sync/pending_blocks_queue.go:89]: we start a new span for each slot/block in the loop, so it makes sense to end it here. The same is done below

File: beacon-chain/sync/validate_aggregate_proof.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/sync/validate_aggregate_proof.go:58]: Yeah that makes sense too, although it isnt part of the spec. Would be a logical check to add too over here.

**Pull request merged**